portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -27,31 +32,45 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/comments")
 public class DataServlet extends HttpServlet {
-  private Comments comments = new Comments();
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
+  // Retrieves all comments and returns it as a JSON string, sorted in descending timestamp { comments: [...] }
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query(Constants.KIND_COMMENT)
+            .addSort(Constants.PROPERTY_TIMESTAMP, Query.SortDirection.DESCENDING);
+    PreparedQuery preparedQuery = datastore.prepare(query);
+
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity: preparedQuery.asIterable()) {
+      String content = (String) entity.getProperty(Constants.PROPERTY_CONTENT);
+      long timestamp = (long) entity.getProperty(Constants.PROPERTY_TIMESTAMP);
+      comments.add(new Comment(content, timestamp));
+    }
+
     String json = new Gson().toJson(comments);
 
     response.setContentType(Constants.TYPE_JSON);
     response.getWriter().println(json);
   }
 
+  // adds a comment entity to the datastore together with timestamp, provided the comment is non-empty.
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String comment = extractComment(request);
+    String content = request.getParameter(Constants.FORM_COMMENT);
+    long timestamp = System.currentTimeMillis();
+
+    Entity commentEntity = new Entity(Constants.KIND_COMMENT);
+    commentEntity.setProperty(Constants.PROPERTY_CONTENT, content);
+    commentEntity.setProperty(Constants.PROPERTY_TIMESTAMP, timestamp);
 
     // don't add empty (useless) comments

```

Comment:
Nit: consider early return to avoid declare unnecessary variables (e.g. commentEntity and timestamp here) in the memory stack.